packages/next/src/client/components/navigation.ts

```diff
@@ -69,6 +70,18 @@ export function useSearchParams(): ReadonlyURLSearchParams {
     return new ReadonlyURLSearchParams(searchParams)
   }, [searchParams]) as ReadonlyURLSearchParams
 
+  // Instrument with Suspense DevTools (dev-only)
+
+  if (process.env.NODE_ENV !== 'production') {
+    // This hook is in a conditional but that is ok because `process.env.NODE_ENV` never changes
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    const navigationPromises = useContext(NavigationPromisesContext)
+    if (navigationPromises) {
+      navigationPromises.searchParams.value = readonlySearchParams
```

Comment:
It seems wrong to mutate this far down.

Wouldn’t you know the context further up? Even if you didn’t, then you should pass a promise that resolves when you do.

Also is it possible that this observes different values in different subtrees and therefore overrides the shared promise with the wrong value?

Reply:
Yeah it's trivial to do it further up for things like pathname, params, searchParams. The trickier one is `useSelectedLayoutSegment` where you can pass in any arbitrary `parallelRouteKey`. I think to achieve that we'd need to instrument promises for every segment of the tree, which seems like it'll get fairly complex.